lib/node/abomination.js

```diff
@@ -1,41 +1,46 @@
-// This abomination solves the empty file extension issue caused by --experimental-loader.
-// Indeed, this flag (seems to) cause the default loader to go from cjs to esm.
-// And file without extension works fine with the cjs loader but not with the esm loader.
+const Module = require("module");
+const { copyFileSync, readFileSync, realpathSync } = require("fs");
+const {
+  extname: getExtension,
+  dirname: getDirectory,
+  basename: getBasename,
+  join: joinPath,
+} = require("path");
 
-const { copyFileSync, readlinkSync, symlinkSync, unlinkSync } = require("fs");
-const { dirname, extname, resolve } = require("path");
+// https://github.com/nodejs/node/blob/2cc7a91a5d855b4ff78f21f1bb8d4e55131d0615/lib/internal/modules/run_main.js
 
-const { stringify } = JSON;
+const { runMain } = Module;
 
-const { argv, stdout } = process;
-const link = argv[1];
-
-let path;
-try {
-  path = readlinkSync(link);
-} catch (error) {
-  path = null;
-}
+const readFileMaybe = (path) => {
+  try {
+    return readFileSync(path, "utf8");
+  } catch (error) {
+    if (error.code === "ENOENT") {
+      return null;
+    } else {
+      throw error;
+    }
+  }
+};
 
-if (path !== null) {
-  const extension = extname(path);
-  if (extension === "") {
-    stdout.write(
-      `[AppmapAgent] Symbolic link ${stringify(link)} refers to ${stringify(
-        path,
-      )} which has no extension, this is problematic for the node's esm loader which is enabled by --experimental-loader. To solve this issue, the agent is going to create copy of that file with a '.cjs' extension and overwrite the link to refer to that file. We hope this is okay with you...${"\n"}`,
-      "utf8",
-    );
-    const absolute_path = resolve(dirname(link), path);
-    try {
-      copyFileSync(absolute_path, `${absolute_path}.cjs`);
-      unlinkSync(link);
-      symlinkSync(`${path}.cjs`, link);
-    } catch ({ message }) {
-      stdout.write(
-        `[AppmapAgent] Something went wrong wen resolving the missing file extension issue: ${message}${"\n"}`,
-        "utf8",
+Module.runMain = function executeUserEntryPoint(main = process.argv[1]) {
+  // TODO this prevents the original runMain to suppport --preserve-symlinks-main
+  const path = realpathSync(main);
+  const directory = getDirectory(path);
+  const extension = getExtension(path);
+  const basename = getBasename(path, extension);
+  if (extension === "" || extension === ".node") {
+    const modified_path = joinPath(directory, `${basename}.cjs`);
+    const maybe_content = readFileMaybe(modified_path);
+    if (maybe_content === null) {
+      copyFileSync(path, modified_path);
```

Comment:
I don't know if it really matters, but is worth checking to see the file exists before copying it again?

Reply:
The `readFileMaybe` function returns null if the file does not exists. So the check is already there :)